### PR TITLE
Every feature has its page

### PR DESCRIPTION
**Pull request by zalun**
This is a first step for https://github.com/mozilla/platatus/issues/260
The next one would be to automatically create the embedded script.

### CODE REVIEW

**Comment by zalun** [comment]
r?

File: tests/nodeUnit.js [hunk 1]
```diff
@@ -69,7 +69,45 @@ define(function(require) {
           'dist/public/index.html',
           'dist/public/manifest.json',
           'dist/public/offline-worker.js',
-          'dist/public/status.json'
+          'dist/public/status.json',
+          'dist/public/app-manifest.html',
```
Comment (marco-c) [inline tests/nodeUnit.js]: We should generate this list automatically.

File: src/tpl/feature.html [hunk 2]
```diff
@@ -0,0 +1,116 @@
+{{#*inline "browserStatus"}}
```
Comment (marco-c) [inline src/tpl/feature.html]: Can we somehow share the code between the two templates? Or are they totally different?

File: tests/nodeUnit.js (hunk 1 above)
Comment (digitarald) [inline tests/nodeUnit.js]: Or remove the test? Any concerns there @TimAbraldes ?

File: src/tpl/feature.html (hunk 2 above)
Comment (zalun) [inline src/tpl/feature.html]: Not totally (but slightly different) and that's my goal for tomorrow

File: tests/nodeUnit.js (hunk 1 above)
Comment (TimAbraldes) [inline tests/nodeUnit.js]: It's a trade-off. Is the maintenance burden of the test worse than the chance that we might accidentally land a patch that causes some resources to 404?

It's a pretty burdensome test, and when our app was a single page we might have gotten the same test coverage out of a quick glance at the output page. Now that we have a bunch of separate pages it seems like the test is more valuable than before.

Comment (digitarald) [inline tests/nodeUnit.js]: Having the right output is a good test result, but it either could check at
a different level or use the list of feature .md files as source of truth
and not a hard coded list that will be hard to maintain.

On Wednesday, January 13, 2016, Tim Abraldes <notifications@github.com>
wrote:

> In tests/nodeUnit.js
> <https://github.com/mozilla/platatus/pull/265#discussion_r49663819>:
>
> > @@ -69,7 +69,45 @@ define(function(require) {
> >            'dist/public/index.html',
> >            'dist/public/manifest.json',
> >            'dist/public/offline-worker.js',
> > -          'dist/public/status.json'
> > +          'dist/public/status.json',
> > +          'dist/public/app-manifest.html',
>
> It's a trade-off. Is the maintenance burden of the test worse than the
> chance that we might accidentally land a patch that causes some resources
> to 404?
>
> It's a pretty burdensome test, and when our app was a single page we might
> have gotten the same test coverage out of a quick glance at the output
> page. Now that we have a bunch of separate pages it seems like the test is
> more valuable than before.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/platatus/pull/265/files#r49663819>.
>


-- 
Web Craftsman | harald@mozilla.com | @digitarald [skype/irc/line] |
+1-650-963-6612 [wapp/telegram]

Comment (TimAbraldes) [inline tests/nodeUnit.js]: Yup, +1 on making this list auto-generated. I can add that to my list of TODOs and get to it before the end of next week.

File: src/tpl/feature.html [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  <meta property="og:description" content="Firefox Platform Status provides an implementation and standardization roadmap for web platform features.">
+  <meta property="og:url" content="https://platatus.herokuapp.com/">
+  <meta property="og:site_name" content="Firefox Platform Status">
+  <meta property="og:image" content="https://platatus.herokuapp.com/images/favicon-196.png">
+  <meta name="twitter:card" content="summary_large_image">
+  <meta name="twitter:description" content="Firefox Platform Status provides an implementation and standardization roadmap for web platform features.">
+  <meta name="twitter:title" content="Firefox Platform Status">
+  <meta name="twitter:site" content="@firefox">
+  <meta name="twitter:image" content="https://platatus.herokuapp.com/images/favicon-196.png">
+  <meta name="twitter:creator" content="@firefox">
```
Comment (zalun) [inline src/tpl/feature.html]: Are there any entries specific for the main page only? There might not be needed in feature.html

**Comment by zalun** [comment]
Can we merge it?

File: engine/index.js [hunk 4]
```diff
@@ -421,13 +421,55 @@ handlebars.registerHelper('if_eq', function comparison(left, right, opts) { // N
   return opts.inverse(this);
 });
 
+// status partial
+const featureStatusContents = fs.readFileSync('src/tpl/featureStatusPartial.html', {
+  encoding: 'utf-8',
+});
+// links partial is needed as the feature page is build in slightly different
```
Comment (marco-c) [inline engine/index.js]: Nit: is built in a

File: engine/index.js [hunk 5]
```diff
@@ -421,13 +421,55 @@ handlebars.registerHelper('if_eq', function comparison(left, right, opts) { // N
...
+// way (no summary), we might also switch off links when embedded
+const featureLinksContents = fs.readFileSync('src/tpl/featureLinksPartial.html', {
+  encoding: 'utf-8',
+});
+
 function buildIndex(status) {
   const templateContents = fs.readFileSync('src/tpl/index.html', {
     encoding: 'utf-8',
   });
+  handlebars.registerHelper('featureStatusName', function featureStatusName() {
```
Comment (marco-c) [inline engine/index.js]: Nit: an arrow function would be more readable:
```JavaScript
handlebars.registerHelper('featureStatusName', () => this.slug + '-status');
```

File: engine/index.js [hunk 6]
```diff
@@ -421,13 +421,55 @@ handlebars.registerHelper('if_eq', function comparison(left, right, opts) { // N
...
+    return this.slug + '-status';
+  });
+  handlebars.registerHelper('featureLinksName', function featureLinksName() {
```
Comment (marco-c) [inline engine/index.js]: Nit: here as well

File: engine/index.js [hunk 7]
```diff
@@ -421,13 +421,55 @@ handlebars.registerHelper('if_eq', function comparison(left, right, opts) { // N
...
+  status.features.forEach((featureData) => {
+    // register partials for each feature
+    handlebars.registerPartial(
+        featureData.slug + '-status',
+        handlebars.compile(featureStatusContents)(featureData));
+    handlebars.registerPartial(
+        featureData.slug + '-links',
+        handlebars.compile(featureLinksContents)(featureData));
+  });
   return Promise.resolve(handlebars.compile(templateContents)(status));
```
Comment (marco-c) [inline engine/index.js]: Do we need to wait for handlebars to compile the partials before compiling the template?

File: engine/index.js [hunk 8]
```diff
@@ -421,13 +421,55 @@ handlebars.registerHelper('if_eq', function comparison(left, right, opts) { // N
...
+    handlebars.registerPartial('featureStatus',
+        handlebars.compile(featureStatusContents)(feature));
+    handlebars.registerPartial('featureLinks',
+        handlebars.compile(featureLinksContents)(feature));
+    return {
+      slug: feature.slug,
+      contents: handlebars.compile(templateContents)(feature),
+    };
+  });
+  return Promise.all(promises);
```
Comment (marco-c) [inline engine/index.js]: `promises` here isn't an array of promises

File: src/tpl/feature.html [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+</head>
+<body>
+  <article class="row">
+    <div class="row-inner">
+      <ul id="features">
+        <li class="feature" id="{{slug}}" tabindex="0">
+          <a href="#{{slug}}"><i aria-hidden="true" class="feature-anchor"></i></a>
```
Comment (marco-c) [inline src/tpl/feature.html]: No need for the anchor in the feature page.

File: engine/index.js (hunk 5 above)
Comment (zalun) [inline engine/index.js]: `this` references to the outer scope (not function's) in arrow functions.

Comment (marco-c) [inline engine/index.js]: Ah, I didn't notice the function was bound by `handlebars`. Forget my comment then, let's just make the functions anonymous.

File: engine/index.js (hunk 7 above)
Comment (digitarald) [inline engine/index.js]: Handlebars is not async, if that is the question

Comment (marco-c) [inline engine/index.js]: The question was related to an earlier version of the PR.

**Pull request merged**